api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -664,14 +664,58 @@ public static String capitalizeFirstCharacterOnly(String text) {
     return StringUtils.capitalize(text.toLowerCase());
   }
 
+  private static String generateSqlWithEnvironmentVariables(
+      String query, KernelTypeEnum kernelTypeEnum) {
+    switch (kernelTypeEnum) {
+      case PYTHON:
+        return query.replaceAll(
+            "\\$\\{projectId}.\\$\\{dataSetId}", "\"\"\" + os.environ[\"WORKSPACE_CDR\"] + \"\"\"");
+      case R:
+        return query.replaceAll(
+            "\\$\\{projectId}.\\$\\{dataSetId}", "\", Sys.getenv(\"WORKSPACE_CDR\"), \"");
+      default:
+        return query;
+    }
+  }
+
+  private static String fillInQueryParams(
+      String query, Map<String, QueryParameterValue> queryParameterValueMap) {
+    StringBuilder stringBuilder = new StringBuilder(query);
+    queryParameterValueMap.forEach(
+        (key, value) -> {
+          if (StandardSQLTypeName.ARRAY.equals(value.getType())) {
+            String stringToReplace = "unnest(@".concat(key.concat(")"));
+            int startingIndex = stringBuilder.indexOf(stringToReplace);
+            stringBuilder.replace(
+                startingIndex,
+                startingIndex + stringToReplace.length(),
+                "("
+                    .concat(
+                        nullableListToEmpty(value.getArrayValues()).stream()
+                            .map(QueryParameterValue::getValue)
+                            .collect(Collectors.joining(", ")))
+                    .concat(")"));
+          } else {
+            String stringToReplace = "@".concat(key);
```

Comment:
Is this going to handle strings vs numbers properly? Wouldn't strings need to be quoted?

Reply:
I think the key is always @key_name, so this should work?

Reply:
Sorry, I should have highlighted the code block. I meant to refer to the value with this comment, not the key, i.e. line 704

Reply:
I will take a look at that.

Reply:
What was the outcome here?